R/extract.R

```diff
@@ -67,15 +72,22 @@ filter_chunk_end_positions <- function(starts, ends) {
   if (length_difference == 0 && all(ends > starts)) {
     return(ends)
   }
-  if (length_difference < 0) {
-    stop("Malformed file!", call. = FALSE)
-  }
 
   positions <- sort(c(starts = starts, ends = ends))
   code_start_indexes <- grep("starts", names(positions))
-  code_ends <- positions[1 + code_start_indexes]
 
-  stopifnot(all(grepl("ends", names(code_ends))))
+  code_ends <- positions[pmin(1 + code_start_indexes, length(positions))]
+
+  bad_end_indexes <- which(grepl("starts", names(code_ends), fixed = TRUE))
+  if (length(bad_end_indexes)) {
+    bad_start_positions <- positions[code_start_indexes[bad_end_indexes]]
+    # This error message is formatted like a parse error
+    stop(sprintf(
```

Comment:
I recommend using `gettextf` here -- `lintr` doesn't do translations (yet) but if it does someday, it'll make it easier.

It's ~basically a drop-in replacement given that we're not doing translation. If we truly wanted to prepare for translation, we'd do `stop(domain=NA, gettextf(..., domain="R-linter"))`.

Not crucial but I think it's an easy change, WDYT?

Reply:
Sounds like a good idea for another issue.
The codebase contains 9 relevant occurences of `stop()`. I think we should migrate all of them at the same time.

Reply:
Sounds good. Filed #577 so as not to forget. Also related: #530